text/0000-hidden-impls.md

```diff
@@ -0,0 +1,669 @@
+}
+```
+
+[`E0277`]: https://doc.rust-lang.org/stable/error-index.html#E0277
+
+in all of these cases, the error [`E0277`] will be emitted. A possible variant
+on `E0277` that highlights the particulars of the problem could be:
+
+```rust
+error[E0277]: the trait bound `Thing: Property` is not satisfied
+ --> src/main.rs:<line>:<column>
+  |
+L |     use_property_1::<Thing>();
+  |     ^^^^^^^^^^^^^^^^^^^^^^^ the trait `Property` is implemented for `Thing`,
+  |                             but the implementation is hidden.
+  |
+note: required by `use_property_1`
+```
+
+### Method calls and E0599
+
+Let's assume that `Property` is in scope, and that we've written:
+
+```rust
+// crate B:
+
+fn main() {
+    let thing = make_a_thing();
+    thing.method();
+}
+```
+
+Let's also assume that there's no other `method` defined in some inherent
+or trait method for `Thing`. In that case, error `E0599` will be emitted,
+with a twist:
+
+```rust
+error[E0599]: no method named `method` found for type `Thing` in the current scope
+  --> src/main.rs:15:7
+   |
+L  | struct Thing;
+   | ------------- method `method` not found for this
+...
+L  |     thing.method();
+   |           ^^^^^^
+   |
+   = help: items from traits can only be used if the trait is implemented and in scope
+   = note: an implementation of `Property`, which defines an item `method`,
+           for `Thing` exists, but is hidden
+```
+
+If instead there does exist some trait implemented for `Thing` which does
+provide the item `method`, then the `hidden_function` lint, adapted with
+language appropriate for methods, will be triggered.
+
+## Lint: `hidden_fn` ⇒ Use UFCS
+
+Consider for a moment that we've written:
+
+```rust
+// crate B:
+
+trait Foo {
+    fn function();
+}
+
+fn main() {
+    use crate_a::Property; // ← assume that `Property::function` exists.
+
+    let x = Thing::function();
+}
+```
+
+Here we want to avoid the future possible breakage of `Property for Thing`
+eventually becoming public. Since the compiler is aware of `Property for Thing`
+and that we've brought `Property` into scope, a warning will be emitted
+in addition to the `unused import` lint, suggesting that you should use
+*uniform function call syntax (UFCS)* instead:
+
+```rust
+warning: the trait bound `Thing: Property` is not satisfied
+ --> src/main.rs:<line>:<column>
+  |
+L |     let x = Thing::function();
+  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ a hidden implementation of trait `Property`,
+  |                                which has a function `function`, exists for `Thing`.
+  |
+  | help: rewrite using UFCS:
+  |
+  |     let x = <Thing as Foo>::function();
+  |
+  = note: #[warn(hidden_fn)] on by default
+```
+
+## Hidden, but coherent
+
+It is crucial to note that `Property for Thing` acts, with respect to coherence,
+as if it were public. That means that crate B, which is a reverse dependency of
+A, may not write:
+
+```rust
+// crate B:
+
+impl Property for Thing { .. }
```

Comment:
I'm confused how this would be allowed anyway as (AIUI) neither `Property` nor `Thing` are defined in crate B, so this impl would be invalid, overlap or no overlap.